### PR TITLE
Update to JupyterLab 2.0

### LINKED ISSUE #68: Update to JupyterLab 2.0
JupyterLab 2.0 is coming soon, and we can start upgrading to it:

- switch to `lumino`
- kernel refactoring: use of `sessionContext`

### PR DESCRIPTION
**Pull request by jtpio**
Fixes #68. 

Migration guide: https://jupyterlab.readthedocs.io/en/latest/developer/extension_migration.html#jupyterlab-1-x-to-2-x-extension-migration-guide

### CODE REVIEW

**Comment by jtpio** [comment]
This builds, but it needs a manual pass to make sure everything behaves as expected.

For example, the Python icon in the launcher card doesn't show anymore:

image

File: advanced/kernel-messaging/README.md [hunk 1]
```diff
@@ -29,28 +29,36 @@ execution result.
 
 ## Initializing and managing a kernel session (`panel.ts`)
 
-Jupyterlab provides a class `ClientSession`
-([see the documentation](https://jupyterlab.github.io/jupyterlab/apputils/classes/clientsession.html))
+Jupyterlab provides a class `SessionContext`
+([see the documentation](https://jupyterlab.github.io/jupyterlab/apputils/classes/sessioncontext.html))
```
Comment (jtpio) [inline advanced/kernel-messaging/README.md]: This link doesn't work (yet). We should check whether it gets updated once 2.0 final is published.

**Comment by jtpio** [comment]
Looks like https://github.com/jtpio/jupyterlab-extension-examples/pull/70/commits/b81943df663c7c181ad17dbea49aa5c1606091b7 fixes the icon issue.

image


A new `LabIcon` must be created to register a new icon with a name, and then be looked up here when rendered as a React element: https://github.com/jupyterlab/jupyterlab/blob/b66bb79116285ccbd0876a6263ac2c0d6de27c33/packages/ui-components/src/icon/labicon.tsx#L218

**Comment by echarles** [comment]
Tested this branch and works well with latest jupyterlab master.

**Comment by jtpio** [comment]
Thanks @echarles for trying it out.

We can check with the latest rc (that includes the icon fix from https://github.com/jupyterlab/jupyterlab/pull/7864), which should make https://github.com/jtpio/jupyterlab-extension-examples/commit/b81943df663c7c181ad17dbea49aa5c1606091b7 not necessary anymore.

**Comment by echarles** [comment]
I have tried with jlab 2.rc2 with and without https://github.com/jtpio/jupyterlab-extension-examples/commit/b81943df663c7c181ad17dbea49aa5c1606091b7

The icons are not displayed. I will experiment a bit and see what is going on here...

Screenshot 2020-02-25 at 14 37 51

**Comment by jtpio** [comment]
And that's from a clean install (`rm yarn.lock node_modules`)?

**Comment by echarles** [comment]
This is the clean I have done before running the tests, but I should do it again as yesterday was a bit messy.

```
find . -name node_modules | xargs rm -fr {} || true
find . -name lib | xargs rm -fr {} || true
find . -name build | xargs rm -fr {} || true
find . -name yarn.lock | xargs rm {} || true
find . -name yarn-error.log | xargs rm {} || true
find . -name tsconfig.tsbuildinfo | xargs rm {} || true
```

**Comment by jtpio** [comment]
Updated to the latest `rc.2` packages. But now tslint seems to be giving errors: https://github.com/jtpio/jupyterlab-extension-examples/pull/70/checks?check_run_id=473012537

**Comment by jtpio** [comment]
@echarles just tried locally in a fresh conda environment, and icons are also missing.

**Comment by echarles** [comment]
Maybe we should @telamonian expert eyes here?

**Comment by jtpio** [comment]
Now using the same pattern as in core JupyterLab:

image

See https://github.com/jtpio/jupyterlab-extension-examples/pull/70/commits/f0de1f9f4bbc91bf0e3a98ab1a03b9a51f61a764.

Also quickly tried the "legacy icon as css class" as mentioned in https://github.com/jupyterlab/jupyterlab/issues/7887#issuecomment-590426912, but without success.

**Comment by jtpio** [comment]
The lint issue mentioned above appears to be related to the latest TypeScript 3.8 version.

Changing it to `~3.7.5` fixes it now, but we should upgrade to `3.8` in a separate PR.

**Comment by telamonian** [comment]
> Also quickly tried the "legacy icon as css class" as mentioned in [jupyterlab/jupyterlab#7887 (comment)](https://github.com/jupyterlab/jupyterlab/issues/7887#issuecomment-590426912), but without success.

@jtpio Turns out that although `iconClass` was getting set up correctly on the DOM, there was no longer any CSS styling directives for how to display background images. That's fixed in #7947

**Comment by jtpio** [comment]
Thanks @telamonian, will check that out!

**Comment by jtpio** [comment]
@echarles I think the examples should now work as before, marking it as ready for review.

cc @fcollonval if you also want to have a look.

**Comment by jtpio** [comment]
We could consolidate the explanations for the `LabIcon` in a separate change, maybe after it has been added to the migration guide.

**Comment by jtpio** [comment]
Updated to the 2.0 packages, since 2.0 final was release yesterday.

This should be good to go.

**Comment by echarles** [comment]
I think not all icons are migrated. 
Screenshot 2020-02-29 at 15 37 12

**Comment by echarles** [comment]
I will branch from this and make a pass on the extension to deal with:

1. Having the menus under one menu - For now, we have plenty of 'Examples' and I would rather see them under a nice `Extension examples` menu.

Screenshot 2020-02-29 at 15 39 19

2. We rely on console to show the result of the actions, but for the user, an alert would be easier.

**Comment by jtpio** [comment]
@echarles I think the other examples just don't have any icon set:

https://github.com/jtpio/jupyterlab-extension-examples/blob/2f87d40d2c3ca09ae02a2653b6d11f7dc9935226/basics/signals/src/index.ts#L67-L71

But it would be nice if they had one (even if it's a placeholder).

**Comment by jtpio** [comment]
> I will branch from this and make a pass on the extension to deal with:
> 
>     1. Having the menus under one menu - For now, we have plenty of 'Examples' and I would rather see them under a nice `Extension examples` menu.
> 
> 
> Screenshot 2020-02-29 at 15 39 19
> 
>     1. We rely on console to show the result of the actions, but for the user, an alert would be easier.

Sounds good :+1: 

How about merging this one first to make it easier to iterate and follow?

**Comment by echarles** [comment]
Merged. Thx a lot @jtpio for this great step. I will iterate on master and submit something tomorrow.

**Comment by jtpio** [comment]
Thanks!

**Pull request merged**